core/src/main/java/jenkins/security/SuppressionFilter.java

```diff
@@ -0,0 +1,160 @@
+            } else if (request instanceof HttpServletRequest && response instanceof HttpServletResponse) {
+                String errorId = logException((HttpServletRequest) request, e);
+
+                respondWithSimpleErrorPage((HttpServletResponse) response, errorId);
+            }
+        }
+    }
+
+    private void respondWithSimpleErrorPage(HttpServletResponse response, String errorId) throws ServletException {
+        try {
+            response.setStatus(SC_INTERNAL_SERVER_ERROR);
+            response.setContentType("text/html;charset=UTF-8");
+            response.setHeader("Cache-Control", "no-cache,must-revalidate");
+            PrintWriter w;
+            try {
+                w = response.getWriter();
+            } catch (IllegalStateException x) {
+                w = new PrintWriter(new OutputStreamWriter(response.getOutputStream(), StandardCharsets.UTF_8));
+            }
+            w.println("<html><head><title>" + Messages.SuppressionFilter_Title() + "</title><body>");
+            w.println("<p>" + Messages.SuppressionFilter_ContactAdmin(errorId) + "</p>");
+            w.println("</body></html>");
+            w.close();
+        } catch (Error error) {
+            throw error;
+        } catch (Throwable x) {
+            // if we fail to report this error, bail out
+            throw new ServletException(Messages.SuppressionFilter_ContactAdmin(errorId)); // not chaining x since it might contain something
+        }
+    }
+
+    private void throwServletException(Exception e) throws ServletException {
+        // ServletException chaining is messed up. We go extra mile here to really
+        // chain the cause.
+        ServletException servletException = new ServletException("Request processing failed", e);
+        try {
+            if (servletException.getCause() == null) {
+                servletException.initCause(e);
+            }
+        } catch (IllegalStateException ise) {
+            // just in case.
+        }
+
+        throw servletException;
+    }
+
+    /**
+     * Report an error and then generate a unique ID for that error.
+     */
+    private String logException(HttpServletRequest req, Throwable e) {
+        String id = UUID.randomUUID().toString();
+        LOGGER.log(Level.WARNING, "Request processing failed. URI=" + req.getRequestURI() + " clientIP=" + req.getRemoteAddr() + " ErrorID=" + id, e);
+        return id;
+    }
+
+    private static boolean showStackTrace() {
+        return Jenkins.get().hasPermission(Jenkins.ADMINISTER);
+    }
+
+    public void destroy() {
+        // no-op
+    }
+
+    private boolean containsAccessDeniedException(Exception exception) {
+        Throwable currentException = exception;
+        do {
+            if (currentException instanceof AccessDeniedException) {
+                return true;
+            }
+            currentException = currentException.getCause();
+        } while (currentException != null);
```

Comment:
do not do this - you can have circular references as a you can set the cause of an exception to itself. (or something higher up the chain)  (for example see Throwable.printStackTrace() code)